torch/csrc/Storage.cpp

```diff
@@ -84,15 +84,124 @@ PyObject* THPStorage_NewWithStorage(
+  RefcountedDeleterContext& ctx =
+      *reinterpret_cast<RefcountedDeleterContext*>(ctx_);
+  ctx.refcount--;
+  if (ctx.refcount == 0) {
+    ctx.other_ctx = nullptr;
+    delete &ctx;
+  }
+}
+
+// If the storage has a normal unique DataPtr, replace it with a DataPtr
+// that can be shared between multiple StorageImpls
+void maybeReplaceDataPtrWithShared(c10::Storage storage) {
+  c10::DataPtr& data_ptr = storage.mutable_data_ptr();
+
+  if (data_ptr.get_deleter() == &refcounted_deleter) {
+    // Data pointer is already shared
+    return;
+  }
+
+  void* data = data_ptr.get();
+  void* other_ctx = data_ptr.get_context();
+  c10::DeleterFnPtr other_deleter = data_ptr.get_deleter();
+  c10::Device device = data_ptr.device();
+
+  // Release the context of the original DataPtr so that the data doesn't
+  // get deleted when the original DataPtr is replaced
+  data_ptr.release_context();
+
+  RefcountedDeleterContext* refcount_ctx =
+      new RefcountedDeleterContext(other_ctx, other_deleter);
+
+  c10::DataPtr new_data_ptr(
+      data, reinterpret_cast<void*>(refcount_ctx), &refcounted_deleter, device);
+  storage.set_data_ptr(std::move(new_data_ptr));
+}
+
+// Create a new StorageImpl that points to the same data. If the original
+// StorageImpl's DataPtr is not shared, replace it with a shared DataPtr
+c10::Storage newStorageImplWithSharedDataPtr(c10::Storage storage) {
+  maybeReplaceDataPtrWithShared(storage);
+
+  c10::StorageImpl* storage_impl = storage.unsafeGetStorageImpl();
+
+  c10::DataPtr& data_ptr = storage.mutable_data_ptr();
+  c10::DataPtr new_data_ptr(
+      data_ptr.get(),
+      data_ptr.get_context(),
+      data_ptr.get_deleter(),
+      data_ptr.device());
+
+  reinterpret_cast<RefcountedDeleterContext*>(data_ptr.get_context())->incref();
+
+  c10::Storage new_storage = c10::make_intrusive<c10::StorageImpl>(
+      c10::StorageImpl::use_byte_size_t(),
+      storage_impl->nbytes(),
+      storage_impl->allocator(),
+      /*resizable=*/storage_impl->resizable());
+
+  new_storage.set_data_ptr(std::move(new_data_ptr));
+  return new_storage;
+}
+
 // Wraps the c10::Storage with a storage PyObject
 PyObject* THPStorage_Wrap(c10::Storage storage) {
+  c10::StorageImpl* storage_impl = storage.unsafeGetStorageImpl();
   if (c10::impl::HermeticPyObjectTLS::get_state()) {
+    c10::impl::PyObjectSlot* pyobj_slot = storage_impl->pyobj_slot();
+
+    bool has_pyobj =
+        pyobj_slot->check_pyobj(pyobj_slot->pyobj_interpreter()).has_value();
+    bool is_same_interpreter =
+        pyobj_slot->pyobj_interpreter() == getPyInterpreter();
+
+    c10::Storage storage_ = storage;
+
+    if (has_pyobj && !is_same_interpreter) {
+      // If the StorageImpl has a PyObject that is managed by a different
+      // interpreter than the current one, create a new StorageImpl that points
+      // to the same data and then create the Python storage from that
+      storage_ = newStorageImplWithSharedDataPtr(storage);
```

Comment:
you gonna need some locks here

Reply:
You're right. One of the tests is failing because two threads are clobbering each other

Reply:
For some reason, `std::mutex` doesn't seem to do anything for these threads. I created `std::mutex replace_data_ptr_mutex;` at global scope in this file and then did `std::lock_guard<std::mutex> guard(replace_data_ptr_mutex);` before calling `newStorageImplWithSharedDataPtr`. But the race condition still exists and sometimes they end up calling `maybeReplaceDataPtrWithShared` at the same time and both try to replace the data pointer.

Maybe the threads in multipy must be getting set up in some weird way that prevents `std::mutex` from working properly? I'm not sure

Reply:
Oh wait, this must be related to the fact that the address space is different

Reply:
It's all same process, how are the address spaces different 🤔

Reply:
I don't know, but moving it to c10 fixed it